drivers/watchdog/wdt_intel_adsp.c

```diff
@@ -0,0 +1,238 @@
+
+#include <zephyr/drivers/watchdog.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys_clock.h>
+#include <zephyr/math/ilog2.h>
+
+#include "wdt_dw.h"
+#include "wdt_dw_common.h"
+#include "wdt_intel_adsp.h"
+
+LOG_MODULE_REGISTER(wdt_intel_adsp, CONFIG_WDT_LOG_LEVEL);
+
+#define DEV_NODE				DT_INST(0, DT_DRV_COMPAT)
+#define WDT_INTEL_ADSP_INTERRUPT_SUPPORT	DT_NODE_HAS_PROP(DEV_NODE, interrupts)
+#define WDT_INTEL_ADSP_INTERRUPT_ALLOW_RESET	0x40000000
+
+/* Device run time data */
+struct intel_adsp_wdt_dev_data {
+	uint32_t device[CONFIG_MP_MAX_NUM_CPUS];
+	uint32_t config;
+#if WDT_INTEL_ADSP_INTERRUPT_SUPPORT
+	wdt_callback_t callback;
+#endif
+};
+
+/* Device constant configuration parameters */
+struct intel_adsp_wdt_dev_cfg {
+	uint32_t base;
+	uint32_t clk_freq;
+};
+
+static int intel_adsp_wdt_setup(const struct device *dev, uint8_t options)
+{
+	const struct intel_adsp_wdt_dev_cfg *const dev_config = dev->config;
+	struct intel_adsp_wdt_dev_data *const dev_data = dev->data;
+	const bool can_reset = dev_data->config & WDT_INTEL_ADSP_INTERRUPT_ALLOW_RESET;
+	unsigned int i;
+	int ret;
+
+	ret = dw_wdt_check_options(options);
+	if (ret) {
+		return ret;
+	}
+
+	for (i = 0; i < CONFIG_MP_MAX_NUM_CPUS; i++) {
+		ret = dw_wdt_configure(dev_data->device[i], dev_data->config);
+		if (ret) {
+			return ret;
+		}
+
+#if WDT_INTEL_ADSP_INTERRUPT_SUPPORT
+		dw_wdt_response_mode_set(dev_data->device[i], !!dev_data->callback);
+#endif
+		intel_adsp_wdt_reset_set(dev_config->base, i, can_reset);
+	}
+
+	return 0;
+}
+
+static int intel_adsp_wdt_install_timeout(const struct device *dev,
+					  const struct wdt_timeout_cfg *config)
+{
+	const struct intel_adsp_wdt_dev_cfg *const dev_config = dev->config;
+	struct intel_adsp_wdt_dev_data *const dev_data = dev->data;
+	int ret;
+
+#if WDT_INTEL_ADSP_INTERRUPT_SUPPORT
+	dev_data->callback = config->callback;
+#else
+	if (config->callback) {
+		LOG_ERR("Interrupt is not configured, can't set a callback.");
+		return -ENOTSUP;
+	}
+#endif
+
+	ret = dw_wdt_calc_period(dev_data->device[0], dev_config->clk_freq, config,
+				 &dev_data->config);
+	if (ret) {
+		return ret;
+	}
+
+	if (config->flags & WDT_FLAG_RESET_CPU_CORE) {
+		dev_data->config |= WDT_INTEL_ADSP_INTERRUPT_ALLOW_RESET;
+	}
+
+	return 0;
+}
+
+static int intel_adsp_wdt_feed(const struct device *dev, int channel_id)
+{
+	struct intel_adsp_wdt_dev_data *const dev_data = dev->data;
+
+	if (channel_id >= CONFIG_MP_MAX_NUM_CPUS) {
+		return -EINVAL;
+	}
+
+	dw_wdt_counter_restart(dev_data->device[channel_id]);
+	return 0;
+}
+
+#if WDT_INTEL_ADSP_INTERRUPT_SUPPORT
+static void intel_adsp_wdt_isr(const struct device *dev)
+{
+	const struct intel_adsp_wdt_dev_cfg *const dev_config = dev->config;
+	struct intel_adsp_wdt_dev_data *const dev_data = dev->data;
+	const uint32_t cpu = arch_proc_id();
+	const uint32_t base = dev_data->device[cpu];
+
+	if (dw_wdt_interrupt_status_register_get(base)) {
+		dw_wdt_clear_interrupt(base);
```

Comment:
Normal paradigm is to clear/ack interrupts only after they're handled (by the callback, in this case), to prevent hardware from racing against the handler.  Maybe it doesn't matter for a watchdog, but probably best to swap the order here for propriety.